File: Algorithm/Code_Materials/python_source/charpter16_slide/level1/find_length_of_LCIS.py
# coding=utf-8
class FindLengthOfLCIS:
    def findLengthOfLCIS(self, nums):
        ans = 0
        n = len(nums)
        start = 0
        for i in range(n):
            if i > 0 and nums[i] <= nums[i - 1]:
                start = i
            ans = max(ans, i - start + 1)
        return ans

    def findLengthOfLCIS2(self, nums):
        cur_len = 1
        res = 1
        for i in range(len(nums)):
            if nums[i - 1] > nums[i]:
                cur_len = 1
            else:
                cur_len = cur_len + 1
            res = max(cur_len, res)
        return res


if __name__ == '__main__':
    nums = [1, 3, 5, 4, 7]
    findLengthOfLCIS = FindLengthOfLCIS()
    print findLengthOfLCIS.findLengthOfLCIS2(nums)

File: Algorithm/Code_Materials/Java/src/main/java/com/yugutou/charpter16_slide/level1/FindLengthOfLCIS.java
package com.yugutou.charpter16_slide.level1;

public class FindLengthOfLCIS {
    public static void main(String[] args) {
        int[] nums = {1, 3, 5, 4, 7, 8, 9, 2};
        System.out.println(findLengthOfLCIS_1(nums));
        System.out.println(findLengthOfLCIS_2(nums));
    }

    public static int findLengthOfLCIS_1(int[] nums) {
        int left = 0, right = 0;
        int res = 0;
        while (right < nums.length) {
            //右侧的新元素比左侧的小，则重新开始记录left位置
            //该问题本质就是快慢指针，left就是slow指针
            if (right > 0 && nums[right - 1] >= nums[right]) {
                left = right;
            }
            right++;
            res = Math.max(res, right - left);
        }
        return res;
    }

    public static int findLengthOfLCIS_2(int[] nums) {
        int curLen = 1;//当前连续递增区间的长度
        int res = 1;
        for (int i = 1; i < nums.length; i++) {
            if (nums[i - 1] >= nums[i]) {
                //不满足要求，重新开始计数
                curLen = 1;
            } else {
                curLen++;
            }
            res = Math.max(curLen, res);
        }
        return res;
    }
}
